tests/functional/curate/cram/rename/general.t

```diff
@@ -0,0 +1,72 @@
+
+Setup
+
+  $ export AUGUR="${AUGUR:-$TESTDIR/../../../../../bin/augur}"
+
+
+  $ cat >records.ndjson <<~~
+  > {"accession": "record_1", "country": "country_1"}
+  > {"accession": "record_2", "country": "country_2"}
+  > ~~
+
+Test that running without (rename-specific) arguments doesn't modify data
+
+  $ $AUGUR curate rename < <(cat records.ndjson) > out1.ndjson
+
+  $ diff records.ndjson out1.ndjson
+
+Rename "accession" to "strain" -- the order should be preserved (i.e. strain is first column)
+
+  $ $AUGUR curate rename --field-map "accession=strain" < <(cat records.ndjson)
+  {"strain": "record_1", "country": "country_1"}
+  {"strain": "record_2", "country": "country_2"}
+
+
+Rename "accession" to "country" - single error message is reported as we won't overwrite without --force
+and we don't change the data
+
+  $ $AUGUR curate rename --field-map "accession=country" < <(cat records.ndjson) > out2.ndjson
+  WARNING: skipping rename of accession because record already has a field named country.
+
+  $ diff records.ndjson out2.ndjson
+
+Rename "accession" to "country" using --force
+
+  $ $AUGUR curate rename --field-map "accession=country" --force < <(cat records.ndjson)
+  {"country": "record_1"}
+  {"country": "record_2"}
+
+
+There's no difference between a field that's missing and a field that's empty in the current implementation
+
+  $ cat >records.partial.ndjson <<~~
+  > {"accession": "record_1"}
+  > {"accession": "record_2", "country": ""}
+  > {"accession": "record_3", "country": "country_3"}
+  > ~~
+
+# Force renaming "accession" to "country" is relatively straightforward:
+  $ $AUGUR curate rename --field-map "accession=country" --force < <(cat records.partial.ndjson)
+  {"country": "record_1"}
+  {"country": "record_2"}
+  {"country": "record_3"}
+
+  $ $AUGUR curate rename --field-map "accession=country" --force --output-metadata - < <(cat records.partial.ndjson)
+  country
+  record_1
+  record_2
+  record_3
+
+# However if we don't have --force then things become more complex
+  $ $AUGUR curate rename --field-map "accession=country" < <(cat records.partial.ndjson) 2>/dev/null 
+  {"country": "record_1"}
+  {"country": "record_2"}
+  {"accession": "record_3", "country": "country_3"}
+
+# for TSV outputs I don't think it's possible, as we don't know about the existence of the "accession" column until we've written the header
+# the following is what I _think_ the output should be (or should the column order be flipped?)
```

Comment:
@joverlee521 have you come up against this before in `augur curate`? The curate machinery is designed to read and write data line-by-line, which is fine for NDJSON outputs (although still a bit strange in situations where we may change the field order between lines) but for NDJSON input and TSV output it's a real problem as we can't write anything until we know the header and we can't know the header until we have read all the data.

(I mean, I can read all the data into memory within `rename`'s `run()` block, and then return an iterator over the entire in-memory dataset, but that doesn't feel like it fits within your design.)

Reply:
P.S. This is also a problem for TSV input & output - e.g. consider a TSV with "strain" and "id" columns, and if we run with `--field-map id=strain` and _without_ `--force` then we don't know if we should output a TSV with both columns or just "strain" until we have read all the data. NDJSON gets around this by allowing different ~columns in each row.

Reply:
Or is the general idea to design programs within this constraint? In this situation it would mean making `--force`   the behaviour all the time (perhaps with suppressible errors).

Reply:
Hmm, the streamed NDJSON records are expected to have the exact same fields, which is why the [metadata output uses the first record's fields as the column names](https://github.com/nextstrain/augur/blob/f6ee377336ec3813468fe641fa7910c14e54ced3/augur/io/metadata.py#L467-L474).

Reply:
I'm realizing now that there's nothing on the input side that verifies the records have the exact same fields. 
Seems like that is a check that we should add to the centralized `augur curate` input parsing.

Reply:
Great! If this is a constraint you think should be central to `augur curate` then it makes the logic of `rename` much much simpler. I'll keep working on this PR under this assumption.